gitfile/gitfile.go

```diff
@@ -7,13 +7,38 @@ import (
 	"io"
 )
 
-// nolint:deadcode,unused
+type GitFile struct {
+	fetcher TokenFetcher
+}
+
+var gitFile = This()
+
+// GetFileContents is a main entry function allowing to retrieve file content from the SCM provider.
+// It expects three file location parameters, from which the repository URL and path to the file are mandatory,
+// and optional Git reference for the branch/tags/commidIds.
+// Function type parameter is a callback used when user authentication is needed in order to retrieve the file,
+// that function will be called with the URL to OAuth service, where user need to be redirected.
 func GetFileContents(ctx context.Context, repoUrl, filepath, ref string, callback func(url string)) (io.ReadCloser, error) {
+	return gitFile.GetFileContents(ctx, repoUrl, filepath, ref, callback)
+}
 
-	fileUrl, err := Detect(repoUrl, filepath, ref)
+func (g *GitFile) GetFileContents(ctx context.Context, repoUrl, filepath, ref string, callback func(url string)) (io.ReadCloser, error) {
+	headerStruct := BuildAuthHeader(repoUrl, g.fetcher)
+	authHeader := req.HeaderFromStruct(headerStruct)
+	fileUrl, err := detect(repoUrl, filepath, ref, authHeader)
 	if err != nil {
 		return nil, err
 	}
-	r, _ := req.Get(fileUrl, ctx)
-	return io.NopCloser(bytes.NewBuffer(r.Bytes())), nil
+
+	response, _ := req.Get(fileUrl, ctx, authHeader)
+	return io.NopCloser(bytes.NewBuffer(response.Bytes())), nil
+}
+
+func (g *GitFile) SetTokenFetcher(fetcher TokenFetcher) {
```

Comment:
I'm not sure I like this setter pattern.

Reply:
@sparkoo you dont have to call `.This`. It's auto called  inside of the lib. And in 99% of time nobody will change token fetcher. So no reason to make user life more complex.